server,storage: plumb contexts and node/store/replica log tags

RaduBerinde opened the pull request:
- Set nodeID log tag in the base server context. This context is inherited by the
  various server components. The trick is that we don't know the node ID when we
  set up the contexts, so we set a log tag with a dynamic value which gets updated
  asynchronously.
- Add log tags to request contexts in Store, using new functionality to copy
  log tags from one context to another. This pattern allows adding tags this
  particular layer doesn't have direct knowledge of (like the nodeID tag set by
  the server, or any tags potentially passed to NewServer).
- Shorten log tags (`n` for node, `s` for store) and suppress the '=' (it can
  be included explicitly in the tag name if necessary).
- Add range log tags. We reimplement the atomic rangeDesc using an atomic pointer,
  and use the new type as the value for the range log tag.

Sample messages:

```
I160822 19:43:39.073183 storage/replica.go:2121  [n1,s1,r7{/Table/51-/Max}] applying command with..
E160822 19:43:39.073284 storage/replica.go:2156  [n1,s1,r7{/Table/51-/Max}] error executing raft..
```

This is a reboot of #8715. Note that there is more work to be done to plumb contexts to a few other components in storage.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8753)

<!-- Reviewable:end -->

tamird commented:
Reviewed 8 of 8 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPonxEyRKeXzZ4zxrLG:-KPonxEyRKeXzZ4zxrLH:b-v67ba1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/server/server.go#L302)):*
> ```Go
> // Start starts the server on the specified port, starts gossip and
> // initializes the node using the engines from the server's context.
> // The passed context can be used to trace the server startup. It should be
> ```

Why can't the passed context be a child of `s.Ctx()`?

---

*[storage/replica.go, line 177 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp0E7DEe4b6Z8qSmi6:-KPp0E7DEe4b6Z8qSmi7:bodrk8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/storage/replica.go#L177)):*
> ```Go
> }
> 
> type atomicRangeDesc struct {
> ```

why does this need to be an unsafe.Pointer rather than atomic.Value as before? Seems that embedding the atomic.Value would allow you to remove the `store` method.

---

*[util/log/log_context.go, line 122 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-BIn6yMkViL0k-_q:-KPp-BIn6yMkViL0k-_r:b-2vioyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L122)):*
> ```Go
> 	if toLastTag == nil {
> 		// Easy case: we reuse the same log tag list directly.
> 		return context.WithValue(ctx, contextTagKeyType{}, fromLastTag)
> ```

use `withLogTag` to avoid duplication?

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPp-Du0l7RNY5WzMwQa:b-7xkes0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
> ```Go
> 	}
> 
> 	// Make a copy of the logTag list.
> ```

seems like this logic should go in `withLogTag` - instead of setting `newTag.prev`, that function should traverse down to the last tag and set its `prev`.

---

*[util/log/log_context.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPoylFHIZsX1H2No7KL:-KPoylFIjXo824WTo8Ps:bii0yrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L157)):*
> ```Go
> 		return "?"
> 	}
> 	return fmt.Sprintf("%d", atomic.LoadInt64(&dv.value))
> ```

nit: strconv.FormatInt is probably more performant.

---

*[util/log/log_context.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-rr07NrRLnHcV2FV:-KPp-rr1sJ8v0bD-V0-u:bq29bl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/util/log/log_context.go#L149)):*
> ```Go
> }
> 
> // DynamicIntValueUnknown can be used with Set; it makes the value "?".
> ```

move to previous commit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPonxEyRKeXzZ4zxrLG:-KPr6uSgrzptPx-gBXuJ:b-f3zh6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/server/server.go#L302)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why can't the passed context be a child of `s.Ctx()`?
</details>
Because what we track during startup includes things before even instantiating a server (https://github.com/cockroachdb/cockroach/blob/af6e2824349d37ce3fe488c4072e39c9a99ecdca/cli/start.go#L305)

---

*[storage/replica.go, line 177 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp0E7DEe4b6Z8qSmi6:-KPr7BnSujAKmUxLTB3E:b-x6mo3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/storage/replica.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does this need to be an unsafe.Pointer rather than atomic.Value as before? Seems that embedding the atomic.Value would allow you to remove the `store` method.
</details>
It could be either, `atomic.Value` is just more complex (internally) than we need, plus we avoid the type assertion on load.

---

*[util/log/log_context.go, line 122 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-BIn6yMkViL0k-_q:-KPr81KkZLgpx9BPW334:bf9612x) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L122)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> use `withLogTag` to avoid duplication?
</details>
withLogTag has to traverse the context hierarchy again (via Value)

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPr8QAaWKoqN-lXVmn6:bf1ogj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like this logic should go in `withLogTag` - instead of setting `newTag.prev`, that function should traverse down to the last tag and set its `prev`.
</details>
The way the links work is: newer tag (in new context) points to older tag (in ancestor context). I don't understand exactly what you are suggesting but note that we are not allowed to change any `logTag` used by an ancestor context.

---

*[util/log/log_context.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPoylFHIZsX1H2No7KL:-KPr9RLTLLV9dmY_JC9m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: strconv.FormatInt is probably more performant.
</details>
Done.

---

*[util/log/log_context.go, line 149 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-rr07NrRLnHcV2FV:-KPr9dOFSFIGH2JssuHq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/util/log/log_context.go#L149)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> move to previous commit
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[server/server.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPonxEyRKeXzZ4zxrLG:-KPsUmyCF-XdZirYB_-6:b-9zrnbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/server/server.go#L302)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Because what we track during startup includes things before even instantiating a server (https://github.com/cockroachdb/cockroach/blob/af6e2824349d37ce3fe488c4072e39c9a99ecdca/cli/start.go#L305)
</details>
You should add that to the comment.

---

*[storage/replica.go, line 177 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp0E7DEe4b6Z8qSmi6:-KPsU-PvVqD-PKdmoX9X:b55bsug) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/storage/replica.go#L177)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> It could be either, `atomic.Value` is just more complex (internally) than we need, plus we avoid the type assertion on load.
</details>
The downside is we deal with a pointer, rather than a value.

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPsUV3Sji3brvLVCNNT:b-1t01zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The way the links work is: newer tag (in new context) points to older tag (in ancestor context). I don't understand exactly what you are suggesting but note that we are not allowed to change any `logTag` used by an ancestor context.
</details>
I'm suggesting that this traversal + copy logic should live in `withLogTag` - it will degenerate to the current implementation in the common case, but should handle more complex needs like this one correctly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[server/server.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPonxEyRKeXzZ4zxrLG:-KPsZKGTySKSI8hAd8Jn:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/server/server.go#L302)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You should add that to the comment.
</details>
Will do.

---

*[storage/replica.go, line 177 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp0E7DEe4b6Z8qSmi6:-KPsX8wMs3nS2ZkqRqU3:b-fys9i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0199d3f3e4b55b7807cd399691c12c5b9eb8ebbe/storage/replica.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> The downside is we deal with a pointer, rather than a value.
</details>
This was already the case, we were storing a pointer (`*RangeDescriptor`) in the `atomic.Value`

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPsYBGG2ziwM4pgymEV:bq0zo9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'm suggesting that this traversal + copy logic should live in `withLogTag` - it will degenerate to the current implementation in the common case, but should handle more complex needs like this one correctly.
</details>
But in all the other cases we don't want `withLogTag` to make a copy. So we could make it accept a log tag chain (instead of a single entry) and connect it properly, but we would still need this loop here to make a copy of the chain. I don't see much of a benefit at this point but I can do that if you think it's better.

In the future, please try to be more specific about your suggestion when it's just bike shedding. If you point out a correctness issue, then it's up to me to figure it out how to best fix it - but when you're talking about another way of doing the same thing, it's hard to guess what you have in mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I'll dump some stuff we've discussed for the archeologists that are gonna be wondering what this node id shenanigans are about. So the problem is that we want the node ID to be part of a bunch of "base contexts" (for the server, the senders, etc), but the node id is only known after we init the stores in `Server.Start()`, whereas all these contexts are created and even used before that. So overwriting them in `Start()` is racy.
We could get the node id early by manually creating an engine and reading the relevant key, but the problem is that for new nodes, ids are assigned by doing a database txn and incrementing some counter. Maybe we could switch to node-generated UUIDs, but they need to be small because they're passed over the network with RPCs.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[kv/dist_sender.go, line 264 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs4fZe0azSfeSlI7qh:-KPs4fZfLEjA-ZDOxuxk:b-5tqxy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/kv/dist_sender.go#L264)):*
> ```Go
> 	// TODO(tschottdorf): Ideally we would use the trace of the request which
> 	// caused this lookup.
> 	br, err := ds.sendRPC(ds.Ctx, desc.RangeID, replicas, ba)
> ```

how about we wrap this `ds.Ctx` in some sort of a `TODO(ds.Ctx)` since we really should have an operation context here

---

*[server/server.go, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs5SDbgGtq0uaB8chw:-KPs5SDbgGtq0uaB8chx:b-yxvrlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L83)):*
> ```Go
> 	gossip         *gossip.Gossip
> 	storePool      *storage.StorePool
> 	txnCoordSender *kv.TxnCoordSender
> ```

Was there any particular reason you made the `txnCoordSender` a member? I think it can be gotten from `db.GetSender()`. Not that I'm against it, but I think it's weird there's members for both `txnCoordSender` and `distSender`. We probably should keep only one (and that should probably be the `txnCoordSender` cause it's higher level?). Maybe you can clean up a little if you're here. Or add some comments if you figure out that both are needed.
Or not.

---

*[server/server.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsQOaxxZvMEelFqyJ2:-KPsQOayqixC7aIi8lQc:b-jgykoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L102)):*
> ```Go
> 	leaseMgr       *sql.LeaseManager
> 
> 	nodeLogTagVal log.DynamicIntValue
> ```

instead of this DynamicIntValue, we could also create a targeted interface, say `NodeIDProvider`, that the `Server` implements and just put that `Server` in the contexts. Like you already do for the range, I think.

---

*[server/server.go, line 141 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs7a5n3A-YGe58Z4-4:-KPs7a5n3A-YGe58Z4-5:b-qzpxss) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L141)):*
> ```Go
> 	}
> 	// Add a dynamic log tag value for the node ID (which we won't know until we
> 	// Start).
> ```

can you expand this comment a bit? So the problem is that we don't know the node id until we call `Start()`, but we need to pass the server's `context` to other pieces before then, so `Start()`is too late for us to update any contexts, right?
I guess the atomic load from you `DynamicVal` is not costing us anything but still maybe there's 

---

*[server/server.go, line 303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs9ISR_h147YT7mCwW:-KPs9ISR_h147YT7mCwX:b-2elyoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L303)):*
> ```Go
> // initializes the node using the engines from the server's context.
> // The passed context can be used to trace the server startup. It should be
> // based off context.Background (not s.Ctx()).
> ```

I think the reference to `context.Background()` doesn't help. This context shouldn't be based on `Background` any more than any context is based on `Background`, right? Just say that the context needs to represent the general server startup operation, and is different from contexts used at runtime for server's background work.

---

*[storage/store.go, line 664 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsLYRZIgP3S3bIVDvi:-KPsLYRZIgP3S3bIVDvj:bjrw5cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/storage/store.go#L664)):*
> ```Go
> }
> 
> // logContext adds the node and store log tags to a context.
> ```

add "Used to personalize an operation context with this Store's identity."

---

*[util/log/log_context.go, line 119 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsMlmK5XC5cXKRbVZu:-KPsMlmK5XC5cXKRbVZv:b-shuipu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/util/log/log_context.go#L119)):*
> ```Go
> 	}
> 
> 	toLastTag := contextLastTag(ctx)
> ```

maybe take the opportunity to add a comment to `logTag` about how `logTags` maintain an overlay over the chain of Contexts

---

*[util/log/log_context.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsOehhYpIIY5Uk4UMM:-KPsOehhYpIIY5Uk4UMN:b-f4sagc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/util/log/log_context.go#L139)):*
> ```Go
> 		prevPtr = &tCopy.prev
> 	}
> 	*prevPtr = toLastTag
> ```

So you're putting the tags from `fromCtx` "last" (i.e. `contextLastTag()` on the new context will return the `fromLastTag`)? What about deduping by name? Not needed after all?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-:-KPs_igXutgGXQlyLliT:b4zas66)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[kv/dist_sender.go, line 264 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs4fZe0azSfeSlI7qh:-KPsbyPxa45mU1bk2Hwl:bc4lx36) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/kv/dist_sender.go#L264)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about we wrap this `ds.Ctx` in some sort of a `TODO(ds.Ctx)` since we really should have an operation context here
</details>
Seems like the TODO here already captures that intent.

---

*[server/server.go, line 303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs9ISR_h147YT7mCwW:-KPsc8TGkJJY0ZSEDVU2:b-pbdqz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L303)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think the reference to `context.Background()` doesn't help. This context shouldn't be based on `Background` any more than any context is based on `Background`, right? Just say that the context needs to represent the general server startup operation, and is different from contexts used at runtime for server's background work.
</details>
Agreed.

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPscEc5oZmvTaf2Onxr:b-wyojdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
> But in all the other cases we don't want withLogTag to make a copy. So we could make it accept a log tag chain (instead of a single entry) and connect it properly, but we would still need this loop here to make a copy of the chain. I don't see much of a benefit at this point but I can do that if you think it's better.

Good point. However, if you made `withLogTag` take a value rather than a pointer, then you could always make a copy and it'd be just fine since no allocations would be involved.

> In the future, please try to be more specific about your suggestion when it's just bike shedding. If you point out a correctness issue, then it's up to me to figure it out how to best fix it - but when you're talking about another way of doing the same thing, it's hard to guess what you have in mind.

In this case, the suggestion wasn't fully formed, but I'll take that under advisement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
storage: range log tags

Add range log tags. We reimplement the atomic rangeDesc using an atomic pointer,
and use the new type as the value for the range log tag.

Sample messages:
```
I160822 19:43:39.073183 storage/replica.go:2121  [n1,s1,r7{/Table/51-/Max}] applying command with..
E160822 19:43:39.073284 storage/replica.go:2156  [n1,s1,r7{/Table/51-/Max}] error executing raft..
```

Tracked under #7994.

RaduBerinde commented:
Review status: 5 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender.go, line 264 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs4fZe0azSfeSlI7qh:-KPsb_JAS4V3O_mC3RNz:b6tosvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/kv/dist_sender.go#L264)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about we wrap this `ds.Ctx` in some sort of a `TODO(ds.Ctx)` since we really should have an operation context here
</details>
What does the `TODO()` do? Does it add anything to the context, or is it a no-op just to not forget it in the code? (we already have the TODO comment for that..) 

---

*[server/server.go, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs5SDbgGtq0uaB8chw:-KPsbtCGF_lK6v3xX254:b7vhanq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L83)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Was there any particular reason you made the `txnCoordSender` a member? I think it can be gotten from `db.GetSender()`. Not that I'm against it, but I think it's weird there's members for both `txnCoordSender` and `distSender`. We probably should keep only one (and that should probably be the `txnCoordSender` cause it's higher level?). Maybe you can clean up a little if you're here. Or add some comments if you figure out that both are needed.
> Or not.
</details>
It was from an earlier iteration where I was setting the node ID in the txnCoordSender during Start(). I figured why not leave it in case it's needed but I can remove. I think the `distSender` is used directly in some tests so it's probably better to leave it.

---

*[server/server.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsQOaxxZvMEelFqyJ2:-KPscHsVogvN4oDnZoeD:b6di9lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L102)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> instead of this DynamicIntValue, we could also create a targeted interface, say `NodeIDProvider`, that the `Server` implements and just put that `Server` in the contexts. Like you already do for the range, I think.
</details>
The implementation of `NodeIDProvider` is basically identical to `DynamicIntValue`, I figured why not make it reusable.

---

*[server/server.go, line 141 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs7a5n3A-YGe58Z4-4:-KPsh8U3hPpWFp-2P0Xv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L141)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you expand this comment a bit? So the problem is that we don't know the node id until we call `Start()`, but we need to pass the server's `context` to other pieces before then, so `Start()`is too late for us to update any contexts, right?
> I guess the atomic load from you `DynamicVal` is not costing us anything but still maybe there's 
</details>
Done.

---

*[server/server.go, line 303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs9ISR_h147YT7mCwW:-KPscpRm0YfChgPLogIA:b-oq3912) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L303)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Agreed.
</details>
Done, PTAL

---

*[storage/store.go, line 664 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsLYRZIgP3S3bIVDvi:-KPshIrbn6sdIHV7IYsb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/storage/store.go#L664)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> add "Used to personalize an operation context with this Store's identity."
</details>
Done.

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPsdW1kZZ1pXptVOZOK:bqg4qm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > But in all the other cases we don't want withLogTag to make a copy. So we could make it accept a log tag chain (instead of a single entry) and connect it properly, but we would still need this loop here to make a copy of the chain. I don't see much of a benefit at this point but I can do that if you think it's better.
> 
> Good point. However, if you made `withLogTag` take a value rather than a pointer, then you could always make a copy and it'd be just fine since no allocations would be involved.
> 
> > In the future, please try to be more specific about your suggestion when it's just bike shedding. If you point out a correctness issue, then it's up to me to figure it out how to best fix it - but when you're talking about another way of doing the same thing, it's hard to guess what you have in mind.
> 
> In this case, the suggestion wasn't fully formed, but I'll take that under advisement.
</details>
The suggestion not being fully formed *is* the problem :)

So you want `withLogTag` to take a `logTag` value, and yet support a *chain*, i.e. other `logTag`s that the value points to. And from what you say it wouldn't make a copy of those in the chain. So the semantics are that you pass a value and that entry (evidently) gets copied, but the rest of the chain is used without copying. That kind of semantic is gross IMO - the function interface suggests it's making a copy of whatever it's passed and yet it doesn't quite do that.

---

*[util/log/log_context.go, line 119 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsMlmK5XC5cXKRbVZu:-KPsjJ3Fa1-R2FD4Ehvi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/util/log/log_context.go#L119)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> maybe take the opportunity to add a comment to `logTag` about how `logTags` maintain an overlay over the chain of Contexts
</details>
Done.

---

*[util/log/log_context.go, line 139 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsOehhYpIIY5Uk4UMM:-KPseyiGVnqkXR2VbbDP:bubv9ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/util/log/log_context.go#L139)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> So you're putting the tags from `fromCtx` "last" (i.e. `contextLastTag()` on the new context will return the `fromLastTag`)? What about deduping by name? Not needed after all?
</details>
Yeah, I'm basically appending the tags from the other context (it's equivalent to "for each tag in fromCtx, ctx =  WithLogTag(ctx, ..)"). I updated the function comment to say that.

The deduplication is not needed in this change AFAICT, I was wondering what the correct behavior would be going forward. I'll leave a TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 8 files at r5, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPsofYQei4ndFW5PlcQ:b-6kdr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The suggestion not being fully formed *is* the problem :)
> 
> So you want `withLogTag` to take a `logTag` value, and yet support a *chain*, i.e. other `logTag`s that the value points to. And from what you say it wouldn't make a copy of those in the chain. So the semantics are that you pass a value and that entry (evidently) gets copied, but the rest of the chain is used without copying. That kind of semantic is gross IMO - the function interface suggests it's making a copy of whatever it's passed and yet it doesn't quite do that.
</details>
Isn't that pattern the classic singly linked list pattern? Why copy everything when you only need to copy the splice point?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[util/log/log_context.go, line 125 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPp-PAjRliYYftFIZAt:-KPspBpLj8bzX6FO6H6-:bgqmxm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb23ac10211ed6ccc701e0d48a297a88b2f3f5c2/util/log/log_context.go#L125)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Isn't that pattern the classic singly linked list pattern? Why copy everything when you only need to copy the splice point?
</details>
There is no pattern like that I can think of. Even if there was, the entry we are copying is not a splice point, it is a new entry. And the rest of the chain that we are not copying, we are *modifying* (specifically the entry at the other end of the chain would be modified to point to the last existing entry in the context). The operation you are suggesting is basically "take a chain A and connect it to an existing chain B, but for some strange reason make a copy of (just) the first entry in chain A while you're at it".

We are wasting too much time talking about how to move around 10 lines of code. I normally accept suggestions even if I'm partial or don't fully agree since it's easier than to have this kind of conversation, but here I don't see any alternative along the lines of what you are saying that is reasonable. If you want to discuss more let's do it over IM since it's faster.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 7 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-:-KPtRyv34L9pvRXdxzyb:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
still LGTM

---

Review status: 7 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 264 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs4fZe0azSfeSlI7qh:-KPxDPiDo896nKqKXfYY:b-tce9l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/kv/dist_sender.go#L264)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> What does the `TODO()` do? Does it add anything to the context, or is it a no-op just to not forget it in the code? (we already have the TODO comment for that..) 
</details>
I was thinking the `TODO()` fun would help with grepping through the source, like `Context.TODO` does, since you're crusading against missing contexts. Otherwise, how would you find this guy to fix it? Maybe it can also add an "operation missing" tag to the logs.
No big deal.

---

*[server/server.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsQOaxxZvMEelFqyJ2:-KPxE67RvBiL7NXFZ1in:bjmo0fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L102)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The implementation of `NodeIDProvider` is basically identical to `DynamicIntValue`, I figured why not make it reusable.
</details>
the advantage of the "provider" is that it'd be self-documenting.

---

*[server/server.go, line 303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs9ISR_h147YT7mCwW:-KPxETTYfpp-HrEQgRDM:b-kxb7kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L303)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Done, PTAL
</details>
(missing closing "`" around s.Ctx())

Now I'm just nitpicking, but I don't understand that meaning of "The intention is to allow a startup tracing span to be set up and used before the Server is created". 
I don't see this function taking a context and different from any other function taking a context - the context should represent the ongoing operation and may or may not have a span (and thus be part of a trace). This sentence makes it sound like there's something special about tracing before the Server is started, which is technically false?
The only thing worth explaining  is that this is different from `s.Ctx()`, which you do before.

Anyway, whatever you think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-:-KPxFl1n5k3LxlcArLGC:bh5f435)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 5 of 13 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender.go, line 264 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs4fZe0azSfeSlI7qh:-KPxRA2YcmFdzPwpZXbv:b-ob4lfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/kv/dist_sender.go#L264)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I was thinking the `TODO()` fun would help with grepping through the source, like `Context.TODO` does, since you're crusading against missing contexts. Otherwise, how would you find this guy to fix it? Maybe it can also add an "operation missing" tag to the logs.
> No big deal.
</details>
Added a `_ = context.TODO()` :)

---

*[server/server.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPsQOaxxZvMEelFqyJ2:-KPxROa6kCetXBZgKSVz:be2k40i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L102)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> the advantage of the "provider" is that it'd be self-documenting.
</details>
Isn't the `nodeLogTagVal` variable name enough?

---

*[server/server.go, line 303 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753#-KPs9ISR_h147YT7mCwW:-KPxSthBEQV18g_wqAdV:b-kivml8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acd7b17e9eebdba330258cc259826ae9d115428/server/server.go#L303)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> (missing closing "`" around s.Ctx())
> 
> Now I'm just nitpicking, but I don't understand that meaning of "The intention is to allow a startup tracing span to be set up and used before the Server is created". 
> I don't see this function taking a context and different from any other function taking a context - the context should represent the ongoing operation and may or may not have a span (and thus be part of a trace). This sentence makes it sound like there's something special about tracing before the Server is started, which is technically false?
> The only thing worth explaining  is that this is different from `s.Ctx()`, which you do before.
> 
> Anyway, whatever you think.
</details>
Makes sense, removed that part.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8753)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
server,storage: set node ID log tags

 - Set nodeID log tag in the base server context. This context is inherited by the
various server components. The trick is that we don't know the node ID when we
set up the contexts, so we set a log tag with a dynamic value which gets updated
asynchronously.

 - Add log tags to request contexts in Store, using new functionality to copy
log tags from one context to another. This pattern allows adding tags this
particular layer doesn't have direct knowledge of (like the nodeID tag set by
the server, or any tags potentially passed to NewServer).

 - Shorten log tags (`n` for node, `s` for store) and suppress the '=' (it can
be included explicitly in the tag name if necessary).

Tracked under #7993, #7994.
